diff --git a/sys/kern/uipc_ktls.c b/sys/kern/uipc_ktls.c
index 61d1a171c46..9842be031b5 100644
--- a/sys/kern/uipc_ktls.c
+++ b/sys/kern/uipc_ktls.c
@@ -302,35 +302,20 @@ static void ktls_reset_receive_tag(void *context, int pending);
 static void ktls_reset_send_tag(void *context, int pending);
 static void ktls_work_thread(void *ctx);
 
+bool
+ktls_offload_enabled(void)
+{
+	return ktls_offload_enable;
+}
+
 int
-ktls_copyin_tls_enable(struct sockopt *sopt, struct tls_enable *tls)
+__ktls_copyin_tls_enable(struct sockopt *sopt, struct tls_enable *tls)
 {
-	struct tls_enable_v0 tls_v0;
+	KASSERT(tls != NULL, ("%s: tls must be valid", __func__));
+
 	int error;
 	uint8_t *cipher_key = NULL, *iv = NULL, *auth_key = NULL;
 
-	if (sopt->sopt_valsize == sizeof(tls_v0)) {
-		error = sooptcopyin(sopt, &tls_v0, sizeof(tls_v0), sizeof(tls_v0));
-		if (error != 0)
-			goto done;
-		memset(tls, 0, sizeof(*tls));
-		tls->cipher_key = tls_v0.cipher_key;
-		tls->iv = tls_v0.iv;
-		tls->auth_key = tls_v0.auth_key;
-		tls->cipher_algorithm = tls_v0.cipher_algorithm;
-		tls->cipher_key_len = tls_v0.cipher_key_len;
-		tls->iv_len = tls_v0.iv_len;
-		tls->auth_algorithm = tls_v0.auth_algorithm;
-		tls->auth_key_len = tls_v0.auth_key_len;
-		tls->flags = tls_v0.flags;
-		tls->tls_vmajor = tls_v0.tls_vmajor;
-		tls->tls_vminor = tls_v0.tls_vminor;
-	} else
-		error = sooptcopyin(sopt, tls, sizeof(*tls), sizeof(*tls));
-
-	if (error != 0)
-		return (error);
-
 	if (tls->cipher_key_len < 0 || tls->cipher_key_len > TLS_MAX_PARAM_SIZE)
 		return (EINVAL);
 	if (tls->iv_len < 0 || tls->iv_len > sizeof(((struct ktls_session *)NULL)->params.iv))
@@ -393,6 +378,37 @@ ktls_copyin_tls_enable(struct sockopt *sopt, struct tls_enable *tls)
 	return (error);
 }
 
+int
+ktls_copyin_tls_enable(struct sockopt *sopt, struct tls_enable *tls)
+{
+	struct tls_enable_v0 tls_v0;
+	int error;
+
+	if (sopt->sopt_valsize == sizeof(tls_v0)) {
+		error = sooptcopyin(sopt, &tls_v0, sizeof(tls_v0), sizeof(tls_v0));
+		if (error != 0)
+			return (error);
+		memset(tls, 0, sizeof(*tls));
+		tls->cipher_key = tls_v0.cipher_key;
+		tls->iv = tls_v0.iv;
+		tls->auth_key = tls_v0.auth_key;
+		tls->cipher_algorithm = tls_v0.cipher_algorithm;
+		tls->cipher_key_len = tls_v0.cipher_key_len;
+		tls->iv_len = tls_v0.iv_len;
+		tls->auth_algorithm = tls_v0.auth_algorithm;
+		tls->auth_key_len = tls_v0.auth_key_len;
+		tls->flags = tls_v0.flags;
+		tls->tls_vmajor = tls_v0.tls_vmajor;
+		tls->tls_vminor = tls_v0.tls_vminor;
+	} else
+		error = sooptcopyin(sopt, tls, sizeof(*tls), sizeof(*tls));
+
+	if (error != 0)
+		return (error);
+
+	return __ktls_copyin_tls_enable(sopt, tls);
+}
+
 void
 ktls_cleanup_tls_enable(struct tls_enable *tls)
 {
@@ -595,7 +611,7 @@ ktls_start_kthreads(void)
 	return (error);
 }
 
-static int
+int
 ktls_create_session(struct socket *so, struct tls_enable *en,
     struct ktls_session **tlsp, int direction)
 {
@@ -824,7 +840,7 @@ ktls_create_session(struct socket *so, struct tls_enable *en,
 	return (0);
 }
 
-static struct ktls_session *
+struct ktls_session *
 ktls_clone_session(struct ktls_session *tls, int direction)
 {
 	struct ktls_session *tls_new;
@@ -1079,7 +1095,7 @@ ktls_alloc_rcv_tag(struct inpcb *inp, struct ktls_session *tls,
 	return (error);
 }
 
-static int
+int
 ktls_try_ifnet(struct socket *so, struct ktls_session *tls, int direction,
     bool force)
 {
@@ -1122,7 +1138,7 @@ ktls_try_ifnet(struct socket *so, struct ktls_session *tls, int direction,
 	return (error);
 }
 
-static void
+void
 ktls_use_sw(struct ktls_session *tls)
 {
 	tls->mode = TCP_TLS_MODE_SW;
diff --git a/sys/sys/ktls.h b/sys/sys/ktls.h
index 3e3f0b77e4a..cf96dc55ea4 100644
--- a/sys/sys/ktls.h
+++ b/sys/sys/ktls.h
@@ -244,8 +244,10 @@ typedef enum {
 	KTLS_MBUF_CRYPTO_ST_DECRYPTED = -1,
 } ktls_mbuf_crypto_st_t;
 
+bool ktls_offload_enabled(void);
 void ktls_check_rx(struct sockbuf *sb);
 void ktls_cleanup_tls_enable(struct tls_enable *tls);
+int __ktls_copyin_tls_enable(struct sockopt *sopt, struct tls_enable *tls);
 int ktls_copyin_tls_enable(struct sockopt *sopt, struct tls_enable *tls);
 void ktls_disable_ifnet(void *arg);
 int ktls_enable_rx(struct socket *so, struct tls_enable *en);
@@ -268,6 +270,11 @@ bool ktls_pending_rx_info(struct sockbuf *sb, uint64_t *seqnop, size_t *residp);
 bool ktls_permit_empty_frames(struct ktls_session *tls);
 void ktls_seq(struct sockbuf *sb, struct mbuf *m);
 int ktls_set_tx_mode(struct socket *so, int mode);
+int ktls_create_session(struct socket *so, struct tls_enable *en,
+			struct ktls_session **tlsp, int direction);
+int ktls_try_ifnet(struct socket *so, struct ktls_session *tls, int direction, bool force);
+void ktls_use_sw(struct ktls_session *tls);
+struct ktls_session *ktls_clone_session(struct ktls_session *tls, int direction);
 
 static inline struct ktls_session *
 ktls_hold(struct ktls_session *tls)
